ceremonies/src/lib.rs

```diff
@@ -358,147 +372,424 @@ decl_error! {
-					Reputation::UnverifiedReputable
-				{
-					reputables.push(participant);
-				} else if <encointer_communities::Module<T>>::bootstrappers(cid)
-					.contains(&participant)
-				{
-					bootstrappers.push(participant);
-				} else if <Endorsees<T>>::contains_key((cid, cindex), &participant) {
-					endorsees.push(participant)
-				} else {
-					newbies.push(participant);
-				}
+			if let Err(e) = Self::generate_meetup_assignment_params((*cid, cindex)) {
+				error!(
+					target: LOG,
+					"Could not generate meetup assignment params for cid: {:?}. {:?}", cid, e
+				);
 			}
+		}
+		Ok(())
+	}
 
-			// n == amount of valid registrations that fit into the meetups
-			let mut n = bootstrappers.len() + reputables.len();
-			n += min(newbies.len(), n / 2);
-			n += endorsees.len();
+	fn mod_inv(a: i64, module: i64) -> i64 {
+		let mut mn = (module, a);
+		let mut xy = (0, 1);
 
-			if n < 3 {
-				debug!(target: LOG, "no meetups assigned for cid {:?}", cid);
-				continue
+		while mn.1 != 0 {
+			xy = (xy.1, xy.0 - (mn.0 / mn.1) * xy.1);
+			mn = (mn.1, mn.0 % mn.1);
+		}
+
+		while xy.0 < 0 {
+			xy.0 += module;
+		}
+		xy.0
+	}
+
+	fn validate_equal_mapping(
+		num_participants: u64,
+		assignment_params: AssignmentParams,
+		n: u64,
+	) -> bool {
+		if num_participants < 2 {
+			return true
+		}
+
+		let mut meetup_index_count: Vec<u64> = vec![0; n as usize];
+		let meetup_index_count_max =
+			checked_ceil_division(num_participants - assignment_params.m, n).unwrap();
+		for i in assignment_params.m..num_participants {
+			let meetup_index = Self::assignment_fn(i, assignment_params, n).unwrap();
+			meetup_index_count[meetup_index as usize] += 1;
+			if meetup_index_count[meetup_index as usize] > meetup_index_count_max {
+				return false
 			}
+		}
+		true
+	}
+
+	fn get_random_nonzero_group_element(m: u64) -> u64 {
+		let mut random_source =
+			RandomNumberGenerator::<T::Hashing>::new(T::RandomnessSource::random_seed().0);
 
-			// ensure that every meetup has at least one experienced participant
-			n = min(n, (bootstrappers.len() + reputables.len()) * 12);
+		// random number in [1, m-1]
+		(random_source.pick_usize((m - 2) as usize) + 1) as u64
+	}
 
-			// capping the amount a participants prevents assigning more meetups than there are locations.
-			if n > n_locations * 12 {
-				warn!(target: LOG, "Meetup Locations exhausted for cid: {:?}", cid);
-				n = n_locations * 12;
+	fn generate_assignment_function_params(
+		num_participants: u64,
+		num_meetups: u64,
+	) -> AssignmentParams {
+		let max_skips = 200;
+		let m = find_prime_below(num_participants);
+		let mut skip_count = 0;
+		let mut s1 = Self::get_random_nonzero_group_element(m);
+		let mut s2 = Self::get_random_nonzero_group_element(m);
+
+		while skip_count <= max_skips {
+			s1 = Self::get_random_nonzero_group_element(m);
+			s2 = Self::get_random_nonzero_group_element(m);
+			if Self::validate_equal_mapping(
```

Comment:
Randomness here is broken. The randomness source is initialized inside the `get_random_nonzero_group_element` function without. Which, which is always the same for the same block, hence all these subsequent calls `get_random_nonzero_group_element` will yield the same value.